libs/ledger-live-common/src/families/cosmos/js-prepareTransaction.ts

```diff
@@ -50,35 +47,30 @@ export const getEstimatedFees = async (
   account: CosmosAccount,
   transaction: Transaction,
 ): Promise<{ estimatedFees: BigNumber; estimatedGas: BigNumber }> => {
-  const cosmosCurrency = cryptoFactory(account.currency.id);
-  let estimatedGas = new BigNumber(cosmosCurrency.defaultGas);
+  const chainInstance = cryptoFactory(account.currency.id);
+  let estimatedGas = new BigNumber(chainInstance.defaultGas);
 
   const cosmosAPI = new CosmosAPI(account.currency.id);
-  const unsignedPayload: { typeUrl: string; value: any }[] = await buildUnsignedPayloadTransaction(
-    account,
-    transaction,
-  );
-
-  if (unsignedPayload && unsignedPayload.length > 0) {
-    const signature = new Uint8Array(Buffer.from(account.seedIdentifier, "hex"));
-
-    // see https://github.com/cosmos/cosmjs/blob/main/packages/proto-signing/src/pubkey.spec.ts
-    const prefix = new Uint8Array([10, 33]);
-
-    const pubkey = {
-      typeUrl: "/cosmos.crypto.secp256k1.PubKey",
-      value: new Uint8Array([...prefix, ...signature]),
-    };
-
-    const tx_bytes = await postBuildUnsignedPayloadTransaction(
-      account,
-      transaction,
-      pubkey,
-      unsignedPayload,
+  const { protoMsgs } = await txToMessages(account, transaction);
+  const { sequence, pubKeyType, pubKey } = await cosmosAPI.getAccount(account.freshAddress);
+  const signature = new Uint8Array(Buffer.from(account.seedIdentifier, "hex"));
+
+  if (protoMsgs && protoMsgs.length > 0) {
+    const tx_bytes = await buildTransaction(
+      protoMsgs,
+      transaction.memo || "",
+      pubKeyType,
+      pubKey,
+      undefined,
+      undefined,
+      sequence + "",
```

Comment:
Beware, I don't think that works
```
undefined + "" === "undefined"
null + "" === "null"
```

Reply:
Ha, I don't know why I assumed it was a string, but it's better that it's a number.
And you're right that here the typing prevents it, so we're good nevermind.

Just to note that the typing doesn't make it impossible though to assign null or undefined at _runtime_:
https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#strictnullchecks-off

And this can especially happen for data that's coming from outside the app, typically from backends like here.
If some backend API decides to stop providing a field that we're expecting, TS won't protect us from having undefined in the expected field.
Just something to remember, but this risk is already here for 99% of the API implementations of our coins, so no difference here.

Reply:
Yes, it's typed as number as we always treat it like this, it comes from backend as a number, and its default value is a number

Reply:
Backend returns a number
Nano returns a string

buildTransaction behavior requires a Long

For now we do Long.fromString in the method, we take nano output as is, and we convert to String on this line

I feel like it is the least error prone way, wdyt ?

Reply:
Hmm I thought about using `String` as well, but it seems discouraged by the doc itself...
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/String
It creates a `String` instance instead of `string` primitive, which indeed may be weird.

Also apparently `String(undefined) === "undefined"` anyway :sweat_smile:

In the end I think I would go with:
```
sequence ? sequence + "" : "0",
```
Uglier but the safest, I think.

Reply:
Done